expr/func_movingwindow.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/grafana/metrictank/consolidation"
+	"github.com/raintank/dur"
+)
+
+type FuncMovingWindow struct {
+	in           GraphiteFunc
+	windowSize   string
+	fn           string
+	xFilesFactor float64
+
+	shiftOffset uint32
+}
+
+// NewMovingWindowConstructor takes an agg string and returns a constructor function
+func NewMovingWindowConstructor(name string) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncMovingWindow{fn: name}
+	}
+}
+
+func NewMovingWindow() GraphiteFunc {
+	return &FuncMovingWindow{fn: "average", xFilesFactor: 0}
+}
+
+func (s *FuncMovingWindow) Signature() ([]Arg, []Arg) {
+	return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgString{key: "windowSize", val: &s.windowSize, validator: []Validator{IsSignedIntervalString}},
+			ArgString{key: "func", opt: true, val: &s.fn, validator: []Validator{IsConsolFunc}},
+			ArgFloat{key: "xFilesFactor", opt: true, val: &s.xFilesFactor, validator: []Validator{WithinZeroOneInclusiveInterval}},
+		}, []Arg{
+			ArgSeriesList{},
+		}
+}
+
+func (s *FuncMovingWindow) Context(context Context) Context {
+	var err error
+	s.shiftOffset, err = s.getWindowSeconds()
+	if err != nil {
+		// shouldn't happen, validated above
+		log.Warnf("movingWindow: encountered error in getWindowSeconds, %s", err)
+		return context
+	}
+
+	// Adjust to fetch from the shifted context.from with
+	// context.to unchanged
+	context.from -= s.shiftOffset
+	return context
+}
+
+func (s *FuncMovingWindow) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+	aggFuncName := fmt.Sprintf("moving%s", strings.Title(s.fn))
+	formatStr := fmt.Sprintf("%s(%%s,\"%s\")", aggFuncName, s.windowSize)
+
+	newName := func(oldName string) string {
+		return fmt.Sprintf(formatStr, oldName)
+	}
+
+	outputs := make([]models.Series, 0, len(series))
+	for _, serie := range series {
+
+		from, points, err := aggregateOnWindow(serie, aggFunc, s.shiftOffset, s.xFilesFactor)
+		if err != nil {
+			return nil, err
+		}
+
+		serie.Target = newName(serie.Target)
+		serie.QueryPatt = newName(serie.QueryPatt)
+		serie.Tags = serie.CopyTagsWith(aggFuncName, s.windowSize)
+		serie.QueryFrom = from
+		serie.Datapoints = points
+
+		outputs = append(outputs, serie)
+	}
+
+	dataMap.Add(Req{}, outputs...)
+	return outputs, nil
+}
+
+func aggregateOnWindow(serie models.Series, aggFunc batch.AggFunc, interval uint32, xFilesFactor float64) (uint32, []schema.Point, error) {
+	out := pointSlicePool.Get().([]schema.Point)
+
+	numPoints := len(serie.Datapoints)
+	serieStart, serieEnd := serie.QueryFrom, serie.QueryTo
+	queryStart := uint32(serieStart + interval)
+
+	if queryStart < serieStart || queryStart > serieEnd {
+		return 0, nil, errors.NewInternalf("query start %d doesn't lie within serie's intervals [%d,%d] ", queryStart, serieStart, serieEnd)
```

Comment:
huh in graphite you can request from=X to=X+12hr and request a windowsize of 24hr for example
it returns a series. why do we return an error?

Reply:
maybe i'm missing something, because testing it in grafana works fine

Reply:
In your example above `queryStart` is X and `serieStart` is 'X-24hrs' and `serieEnd` is X+12hrs.
This sanity check is to make sure that `querystart` is never calculated to be before `serieStart` and after `serieEnd`. However unlikely it is to occur, we want to check to make sure we error out in that case.

Reply:
I see. well given how we adjust context.from in FuncMovingWindow.Context() (we reduce `from`), serieStart will always be before querystart.  but ok, i have no objection against making sure.I do think these lines of code need some clarification about why they're checking this, what the concern is, how things might break if the condition is not met, etc

Reply:
could be as simple as "querystart needs to be between serieStart and serieEnd for the loop below to work (this should always be the case)" . at least that's my interpretation. maybe you can explain it better.

Reply:
Understood. I'll add a comment explaining.